scripts/src/java/org/oppia/android/scripts/todo/TodoCheck.kt

```diff
@@ -0,0 +1,302 @@
+ *
+ * @param codeLine the line of code to be checked
+ * @param openIssueList the list of all the open issues of this repository on GitHub
+ * @param correctTodoFormatRegex regex of the correct todo format
+ * @return whether the todo does not corresponds to an open issue
+ */
+private fun checkIfOpenIssueFailure(
+  codeLine: String,
+  openIssueList: List<Issue>,
+  correctTodoFormatRegex: Regex
+): Boolean {
+  if (!correctTodoFormatRegex.containsMatchIn(codeLine)) {
+    return false
+  }
+  val match = correctTodoFormatRegex.find(codeLine)
+  val parsedIssueNumberFromTodo = match?.groupValues?.get(1)
+  return !openIssueList.any { it -> it.issueNumber == parsedIssueNumberFromTodo }
+}
+
+/**
+ * Logs the redundant exemptions.
+ *
+ * @param redundantExemptions list of redundant exemptions
+ * @param todoExemptionTextProtoFilePath the location of the todo exemption textproto file
+ */
+private fun logRedundantExemptions(
+  redundantExemptions: List<Pair<String, Int>>,
+  todoExemptionTextProtoFilePath: String
+) {
+  if (redundantExemptions.isNotEmpty()) {
+    println("Redundant exemptions:")
+    redundantExemptions.sortedWith(compareBy({ it.first }, { it.second })).forEach { exemption ->
+      println("- ${exemption.first}:${exemption.second}")
+    }
+    println(
+      "Please remove them from $todoExemptionTextProtoFilePath.textproto"
+    )
+    println()
+  }
+}
+
+/**
+ * Logs the Todo check failures.
+ *
+ * @param invalidTodos a list of all the invalid TODOs present in the repository. A TODO is
+ *     considered to be invalid if it is poorly formatted or if it does not corresponds to an open
+ *     issue on GitHub.
+ * @param failureMessage the failure message to be logged
+ * @param failureNote the failure note
+ */
+private fun logFailures(invalidTodos: List<Todo>, failureMessage: String, failureNote: String) {
+  if (invalidTodos.isNotEmpty()) {
+    println(failureMessage)
+    invalidTodos.sortedWith(compareBy({ it.filePath }, { it.lineNumber })).forEach {
+      println("- ${it.filePath}:${it.lineNumber}")
+    }
+    println(failureNote)
+    println()
+  }
+}
+
+/**
+ * Retrieves the list of all open issues on GitHub by parsing the JSON file generated by the GitHub
+ * API.
+ *
+ * @param openIssuesJsonFile file containing all the open issues of the repository
+ * @return list of all open issues
+ */
+private fun retrieveOpenIssueList(openIssuesJsonFile: File): List<Issue> {
+  val openIssuesJsonText = openIssuesJsonFile
+    .inputStream()
+    .bufferedReader()
+    .use { it.readText() }
+  val moshi = Moshi.Builder().addLast(KotlinJsonAdapterFactory()).build()
+  val listType = Types.newParameterizedType(List::class.java, Issue::class.java)
+  val adapter: JsonAdapter<List<Issue>> = moshi.adapter(listType)
+  return adapter.fromJson(openIssuesJsonText)
+    ?: throw Exception("Failed to parse $openIssuesJsonFile")
+}
+
+/**
+ * Loads the todo exemptions list from a text proto file.
```

Comment:
Should this be corresponding to a text proto file, instead?

Reply:
Done.